// @react-native-community/cli/build/commands/upgrade/upgrade.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _fs() {
  const data = _interopRequireDefault(require("fs"));

  _fs = function () {
    return data;
  };

  return data;
}

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function () {
    return data;
  };

  return data;
}

function _semver() {
  const data = _interopRequireDefault(require("semver"));

  _semver = function () {
    return data;
  };

  return data;
}

function _execa() {
  const data = _interopRequireDefault(require("execa"));

  _execa = function () {
    return data;
  };

  return data;
}

var _logger = _interopRequireDefault(require("../../tools/logger"));

var _PackageManager = _interopRequireDefault(require("../../tools/PackageManager"));

var _helpers = require("./helpers");

var _legacyUpgrade = _interopRequireDefault(require("./legacyUpgrade"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const rnDiffPurgeUrl = 'https://github.com/react-native-community/rn-diff-purge';

const getLatestRNVersion = async () => {
  _logger.default.info('No version passed. Fetching latest...');

  const {
    stdout
  } = await (0, _execa().default)('npm', ['info', 'react-native', 'version']);
  return stdout;
};

const getRNPeerDeps = async version => {
  const {
    stdout
  } = await (0, _execa().default)('npm', ['info', `react-native@${version}`, 'peerDependencies', '--json']);
  return JSON.parse(stdout);
};

const getPatch = async (currentVersion, newVersion, projectDir) => {
  let patch;
  const rnDiffAppName = 'RnDiffApp';

  const {
    name
  } = require(_path().default.join(projectDir, 'package.json'));

  _logger.default.info(`Fetching diff between v${currentVersion} and v${newVersion}...`);

  try {
    patch = await (0, _helpers.fetch)(`${rnDiffPurgeUrl}/compare/version/${currentVersion}...version/${newVersion}.diff`);
  } catch (error) {
    _logger.default.error(`Failed to fetch diff for react-native@${newVersion}. Maybe it's not released yet?`);

    _logger.default.info('For available releases to diff see: https://github.com/react-native-community/rn-diff-purge#version-changes');

    return null;
  }

  return patch.replace(new RegExp(rnDiffAppName, 'g'), name).replace(new RegExp(rnDiffAppName.toLowerCase(), 'g'), name.toLowerCase());
};

const getVersionToUpgradeTo = async (argv, currentVersion, projectDir) => {
  const newVersion = argv[0] ? _semver().default.valid(argv[0]) || (_semver().default.coerce(argv[0]) ? _semver().default.coerce(argv[0]).version : null) : await getLatestRNVersion();

  if (!newVersion) {
    _logger.default.error(`Provided version "${argv[0]}" is not allowed. Please pass a valid semver version`);

    return null;
  }

  if (currentVersion > newVersion) {
    _logger.default.error(`Trying to upgrade from newer version "${currentVersion}" to older "${newVersion}"`);

    return null;
  }

  if (currentVersion === newVersion) {
    const {
      dependencies: {
        'react-native': version
      }
    } = require(_path().default.join(projectDir, 'package.json'));

    if (_semver().default.satisfies(newVersion, version)) {
      _logger.default.warn(`Specified version "${newVersion}" is already installed in node_modules and it satisfies "${version}" semver range. No need to upgrade`);

      return null;
    }

    _logger.default.error(`Dependency mismatch. Specified version "${newVersion}" is already installed in node_modules and it doesn't satisfy "${version}" semver range of your "react-native" dependency. Please re-install your dependencies`);

    return null;
  }

  return newVersion;
};

const installDeps = async (newVersion, projectDir, patchSuccess) => {
  if (!patchSuccess) {
    _logger.default.warn('Continuing after failure. Most of the files are upgraded but you will need to deal with some conflicts manually');
  }

  _logger.default.info(`Installing react-native@${newVersion} and its peer dependencies...`);

  const peerDeps = await getRNPeerDeps(newVersion);
  const pm = new _PackageManager.default({
    projectDir
  });
  const deps = [`react-native@${newVersion}`, ...Object.keys(peerDeps).map(module => `${module}@${peerDeps[module]}`)];
  await pm.install(deps, {
    silent: true
  });
  await (0, _execa().default)('git', ['add', 'package.json']);

  try {
    await (0, _execa().default)('git', ['add', 'yarn.lock']);
  } catch (error) {// ignore
  }

  try {
    await (0, _execa().default)('git', ['add', 'package-lock.json']);
  } catch (error) {// ignore
  }
};

const applyPatch = async (currentVersion, newVersion, tmpPatchFile) => {
  let filesToExclude = ['package.json'];

  try {
    try {
      const excludes = filesToExclude.map(e => `--exclude=${e}`);
      await (0, _execa().default)('git', ['apply', '--check', tmpPatchFile, ...excludes, '-p2', '--3way']);

      _logger.default.info('Applying diff...');
    } catch (error) {
      filesToExclude = [...filesToExclude, ...error.stderr.split('\n').filter(x => x.includes('does not exist in index')).map(x => x.replace(/^error: (.*): does not exist in index$/, '$1'))].filter(Boolean);

      _logger.default.info(`Applying diff (excluding: ${filesToExclude.join(', ')})...`);
    } finally {
      const excludes = filesToExclude.map(e => `--exclude=${e}`);
      await (0, _execa().default)('git', ['apply', tmpPatchFile, ...excludes, '-p2', '--3way']);
    }
  } catch (error) {
    if (error.stderr) {
      _logger.default.log(`${_chalk().default.dim(error.stderr.trim())}`);
    }

    _logger.default.error('Automatically applying diff failed');

    _logger.default.info(`Here's the diff we tried to apply: ${rnDiffPurgeUrl}/compare/version/${currentVersion}...version/${newVersion}`);

    _logger.default.info(`You may find release notes helpful: https://github.com/facebook/react-native/releases/tag/v${newVersion}`);

    return false;
  }

  return true;
};
/**
 * Upgrade application to a new version of React Native.
 */


async function upgrade(argv, ctx, args) {
  if (args.legacy) {
    return _legacyUpgrade.default.func(argv, ctx);
  }

  const rnDiffGitAddress = 'https://github.com/react-native-community/rn-diff-purge.git';
  const tmpRemote = 'tmp-rn-diff-purge';
  const tmpPatchFile = 'tmp-upgrade-rn.patch';
  const projectDir = ctx.root;

  const {
    version: currentVersion
  } = require(_path().default.join(projectDir, 'node_modules/react-native/package.json'));

  const newVersion = await getVersionToUpgradeTo(argv, currentVersion, projectDir);

  if (!newVersion) {
    return;
  }

  const patch = await getPatch(currentVersion, newVersion, projectDir);

  if (patch === null) {
    return;
  }

  if (patch === '') {
    _logger.default.info('Diff has no changes to apply, proceeding further');

    await installDeps(newVersion, projectDir);

    _logger.default.success(`Upgraded React Native to v${newVersion} 🎉. Now you can review and commit the changes`);

    return;
  }

  let patchSuccess;

  try {
    _fs().default.writeFileSync(tmpPatchFile, patch);

    await (0, _execa().default)('git', ['remote', 'add', tmpRemote, rnDiffGitAddress]);
    await (0, _execa().default)('git', ['fetch', '--no-tags', tmpRemote]);
    patchSuccess = await applyPatch(currentVersion, newVersion, tmpPatchFile);

    if (!patchSuccess) {
      return;
    }
  } catch (error) {
    throw new Error(error.stderr || error);
  } finally {
    try {
      _fs().default.unlinkSync(tmpPatchFile);
    } catch (e) {// ignore
    }

    await installDeps(newVersion, projectDir, patchSuccess);

    _logger.default.info('Running "git status" to check what changed...');

    await (0, _execa().default)('git', ['status'], {
      stdio: 'inherit'
    });
    await (0, _execa().default)('git', ['remote', 'remove', tmpRemote]);

    if (!patchSuccess) {
      _logger.default.warn('Please run "git diff" to review the conflicts and resolve them');

      throw new Error('Upgrade failed. Please see the messages above for details');
    }
  }

  _logger.default.success(`Upgraded React Native to v${newVersion} 🎉. Now you can review and commit the changes`);
}

const upgradeCommand = {
  name: 'upgrade [version]',
  description: "Upgrade your app's template files to the specified or latest npm version using `rn-diff-purge` project. Only valid semver versions are allowed.",
  func: upgrade,
  options: [{
    command: '--legacy',
    description: "Legacy implementation. Upgrade your app's template files to the latest version; run this after " + 'updating the react-native version in your package.json and running npm install'
  }]
};
var _default = upgradeCommand;
exports.default = _default;
