// @react-native-community/cli/build/commands/install/install.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _logger = _interopRequireDefault(require("../../tools/logger"));

var _PackageManager = _interopRequireDefault(require("../../tools/PackageManager"));

var _link = _interopRequireDefault(require("../link/link"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
async function install(args, ctx) {
  const name = args[0];

  _logger.default.info(`Installing "${name}"...`);

  new _PackageManager.default({
    projectDir: ctx.root
  }).install([name]);

  _logger.default.info(`Linking "${name}"...`);

  await _link.default.func([name], ctx, {
    platforms: undefined
  });

  _logger.default.success(`Successfully installed and linked "${name}"`);
}

var _default = {
  func: install,
  description: 'install and link native dependencies',
  name: 'install <packageName>'
};
exports.default = _default;
